lib/libutee/tee_api_objects.c

```diff
@@ -38,24 +38,92 @@
 #define TEE_ATTR_BIT_PROTECTED              (1 << 28)
 
 /* Data and Key Storage API  - Generic Object Functions */
+/*
+ * Use of this function is deprecated
+ * new code SHOULD use the TEE_GetObjectInfo1 function instead
+ * These functions will be removed at some future major revision of
+ * this specification
+ */
 void TEE_GetObjectInfo(TEE_ObjectHandle object, TEE_ObjectInfo *objectInfo)
 {
 	TEE_Result res;
 
 	res = utee_cryp_obj_get_info((uint32_t)object, objectInfo);
+
 	if (res != TEE_SUCCESS)
 		TEE_Panic(res);
+
+	if (objectInfo->objectType == TEE_TYPE_CORRUPTED_OBJECT) {
+		objectInfo->keySize = 0;
+		objectInfo->maxKeySize = 0;
+		objectInfo->objectUsage = 0;
+		objectInfo->dataSize = 0;
+		objectInfo->dataPosition = 0;
+		objectInfo->handleFlags = 0;
+	}
 }
 
+TEE_Result TEE_GetObjectInfo1(TEE_ObjectHandle object, TEE_ObjectInfo *objectInfo)
+{
+	TEE_Result res;
+
+	res = utee_cryp_obj_get_info((uint32_t)object, objectInfo);
+
+	if (res != TEE_SUCCESS && res != TEE_ERROR_CORRUPT_OBJECT &&
+	    res != TEE_ERROR_STORAGE_NOT_AVAILABLE)
+		TEE_Panic(res);
+
+	if (res == TEE_ERROR_CORRUPT_OBJECT) {
+		res = utee_storage_obj_del(object);
+		if (res != TEE_SUCCESS)
+			TEE_Panic(0);
+		return TEE_ERROR_CORRUPT_OBJECT;
+	}
+
+	return res;
+}
+
+/*
+ * Use of this function is deprecated
+ * new code SHOULD use the TEE_RestrictObjectUsage1 function instead
+ * These functions will be removed at some future major revision of
+ * this specification
+ */
 void TEE_RestrictObjectUsage(TEE_ObjectHandle object, uint32_t objectUsage)
 {
 	TEE_Result res;
-	res = utee_cryp_obj_restrict_usage((uint32_t)object, objectUsage);
+	TEE_ObjectInfo objectInfo;
+
+	res = utee_cryp_obj_get_info((uint32_t)object, &objectInfo);
+	if (objectInfo.objectType == TEE_TYPE_CORRUPTED_OBJECT)
+		return;
+
+	res = TEE_RestrictObjectUsage1(object, objectUsage);
 
 	if (res != TEE_SUCCESS)
 		TEE_Panic(0);
 }
 
+TEE_Result TEE_RestrictObjectUsage1(TEE_ObjectHandle object, uint32_t objectUsage)
+{
+	TEE_Result res;
+
+	res = utee_cryp_obj_restrict_usage((uint32_t)object, objectUsage);
+
+	if (res != TEE_SUCCESS && res != TEE_ERROR_CORRUPT_OBJECT &&
+	    res != TEE_ERROR_STORAGE_NOT_AVAILABLE)
+		TEE_Panic(0);
+
+	if (res == TEE_ERROR_CORRUPT_OBJECT) {
```

Comment:
If this test and action was done before the test for panic at line 113-114 above, it would simplify the test for panic.